File: 04-quick-sort/ruby/quick_sort.rb
# frozen_string_literal: true

def quick_sort(arr)
  return arr if arr.size < 2
  pivot_index = rand(arr.size - 1)
  pivot = arr[pivot_index]

  less = []
  greater = []
  arr.size.times do |i|
    next if i == pivot_index
    if arr[i] <= pivot
      less << arr[i]
    else
      greater << arr[i]
    end
  end
  quick_sort(less) + [pivot] + quick_sort(greater)
end

File: 04-quick-sort/python/quick_sort.py
from random import randrange

def quick_sort(lst):
    """Implementation of quick sort algorithm"""
    if len(lst) < 2:
        return lst
    else:
        pivot_index = randrange(len(lst))
        pivot = lst[pivot_index]

        less = []
        greater = []
        for i, x in enumerate(lst):
            if i == pivot_index:
                continue
            if x <= pivot:
                less.append(x)
            else:
                greater.append(x)
        return quick_sort(less) + [pivot] + quick_sort(greater)
